glib-macros/src/downgrade_struct.rs

```diff
@@ -0,0 +1,44 @@
+use crate::downgrade_fields::{derive_downgrade_fields, DowngradeStructParts};
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+use syn::{Generics, Ident};
+
+pub fn derive_downgrade_for_struct(
+    ident: Ident,
+    generics: Generics,
+    data_struct: syn::DataStruct,
+) -> TokenStream {
+    let weak_type = format_ident!("{}Weak", ident);
+
+    let DowngradeStructParts {
+        weak_fields,
+        end_of_struct,
+        destruct,
+        downgrade,
+        upgrade,
+    } = derive_downgrade_fields(data_struct.fields);
+
+    let derived = quote! {
+        pub struct #weak_type #generics #weak_fields #end_of_struct
```

Comment:
It's slightly ugly that we let a new type pop up in the scope as part of the derive macro but I don't think there's anything better we can do here so...

Reply:
Alternative is to generate name like [gensym](https://crates.io/crates/gensym) does. But this will make build non-reproducible. Which is, i believe, unwanted.

Anyway, type leaks through `<T as glib::clone::Downgrade>::Weak`.

Reply:
Yes that's not really an improvement :) Let's keep it like this for now